ndpyramid/utils.py

```diff
@@ -29,3 +51,107 @@ def multiscales_template(
             'metadata': {'method': method, 'version': version, 'args': args, 'kwargs': kwargs},
         }
     ]
+
+
+def set_zarr_encoding(
+    ds: xr.Dataset,
+    codec_config: dict | None = None,
+    float_dtype: npt.DTypeLike | None = None,
+    int_dtype: npt.DTypeLike | None = None,
+) -> xr.Dataset:
+    """Set zarr encoding for each variable in the dataset
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Input dataset
+    codec_config : dict, optional
+        Dictionary of parameters to pass to numcodecs.get_codec.
+        The default is {'id': 'zlib', 'level': 1}
+    float_dtype : str or dtype, optional
+        Dtype to cast floating point variables to
+
+    Returns
+    -------
+    ds : xr.Dataset
+        Output dataset with updated variable encodings
+    """
+    import numcodecs
+
+    ds = ds.copy()
+
+    if codec_config is None:
+        codec_config = {'id': 'zlib', 'level': 1}
+    compressor = numcodecs.get_codec(codec_config)
+
+    for k, da in ds.variables.items():
+
+        # maybe cast float type
+        if np.issubdtype(da.dtype, np.floating) and float_dtype is not None:
+            da = da.astype(float_dtype)
+
+        if np.issubdtype(da.dtype, np.integer) and int_dtype is not None:
+            da = da.astype(int_dtype)
+
+        # remove old encoding
+        da.encoding.clear()
+
+        # update with new encoding
+        da.encoding['compressor'] = compressor
+        with contextlib.suppress(KeyError):
+            del da.attrs['_FillValue']
+        da.encoding['_FillValue'] = default_fillvals.get(
+            da.dtype.str[-2:], None
+        )  # TODO: handle date/time types
+
+        ds[k] = da
+
+    return ds
+
+
+def _add_metadata_and_zarr_encoding(
+    pyramid: dt.DataTree, *, levels: int, other_chunks: dict = None, pixels_per_tile: int = 128
+) -> dt.DataTree:
+
+    '''Postprocess data pyramid. Adds multiscales metadata and sets Zarr encoding
+
+    Parameters
+    ----------
+    pyramid : dt.DataTree
+        Input data pyramid
+    levels : int
+        Number of levels in pyramid
+    other_chunks : dict
+        Chunks for non-spatial dims
+    pixels_per_tile : int
+        Number of pixels per tile
+
+    Returns
+    -------
+    dt.DataTree
+        Updated data pyramid with metadata / encoding set
+    '''
+    chunks = {'x': pixels_per_tile, 'y': pixels_per_tile}
+    if other_chunks is not None:
+        chunks.update(other_chunks)
+
+    for level in range(levels):
+        slevel = str(level)
+        pyramid.ds.attrs['multiscales'][0]['datasets'][level]['pixels_per_tile'] = pixels_per_tile
+
+        # set dataset chunks
+        pyramid[slevel].ds = pyramid[slevel].ds.chunk(chunks)
+        if 'date_str' in pyramid[slevel].ds:
+            pyramid[slevel].ds['date_str'] = pyramid[slevel].ds['date_str'].chunk(-1)
```

Comment:
This should probably live elsewhere. We were using this as a workaround for a specific application. 

```suggestion
```